### PR TITLE
MM-51310 Implement OAuth2 Authentication (reopened)

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

Reopening of https://github.com/mattermost/mattermost-plugin-jira/pull/928

This PR implements the OAuth2 authentication in the Jira Plugin. 

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-51310

#### Screenshots

**Install OAuth2 using setup flow**
Whenever the system admin runs the command `/jira setup`, the setup flow to configure OAuth2 client would look something like this:

Screenshot 2023-04-25 at 12 23 20 PM

Screenshot 2023-04-25 at 12 23 39 PM

Screenshot 2023-04-17 at 4 33 06 PM

Screenshot 2023-04-17 at 4 34 06 PM

### CODE REVIEW

**Comment by DHaussermann** [comment]
I have done a round of testing on this.

The happy path for setting up an oAuth App and connecting user is working however, there is a problem here where part of the subscriptions functionality is no longer working. 

@mickmister had raised a concern here about comment events on subscriptions https://github.com/mattermost/mattermost-plugin-jira/pull/928#issuecomment-1574267767 He was right to be worried. This indeed does not work anymore.

Comments events are part of the webhook config on Jira side.
image

Comment events are selected as part of the subscription on Mattermost side
image

Several issue events are working but none of the 3 comment events are functional 
image

Logs show
```
{
  "caller": "app/plugin_api.go:976",
  "level": "error",
  "msg": "WebhookWorker id: 8, error processing, err: Must be a JIRA Cloud instance, is cloud-oauth",
  "plugin_id": "jira",
  "timestamp": "2023-06-06 21:45:17.621 Z"
}
```

@jasonblais I will post in the Brightscout maintenance channel that this needs another round of development. Unless you have another resource in mind seeing as how Michael will be away.

**Comment by jasonblais** [comment]
Thanks @DHaussermann! Yes, let's ask BrightScout's help with additional development work

File: server/instance_cloud_oauth.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+func (ci *cloudOAuthInstance) GetDisplayDetails() map[string]string {
+	return map[string]string{
+		"Jira Cloud Mattermost Key": ci.MattermostKey,
+	}
+}
+
+func (ci *cloudOAuthInstance) GetUserConnectURL(mattermostUserID string) (string, *http.Cookie, error) {
+	oauthConf := ci.GetOAuthConfig()
+	state := fmt.Sprintf("%s_%s", model.NewId()[0:15], mattermostUserID)
+	url := oauthConf.AuthCodeURL(
```
[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply server/instance_cloud_oauth.go:127]: @esarafianou Added PKCE along with the authorization code in the following PR: https://github.com/mattermost/mattermost-plugin-jira/pull/953

**Comment by raghavaggarwal2308** [comment]
@mickmister @DHaussermann Fixed issue: Comment webhook notification not working in the following PR: https://github.com/mattermost/mattermost-plugin-jira/pull/953

**Review by esarafianou — APPROVED** [review]

**Comment by BlueSky-fur** [comment]
🥳 2 more to go to make work easier again! 😄

**Comment by hanzei** [comment]
@mickmister What are the next steps on this PR? Is it ready for QA testing?

**Comment by mickmister** [comment]
@hanzei Yes it's ready for QA testing. Discussing in Brightscout channel next steps for testing https://community.mattermost.com/core/pl/ng4tz8stp7ynucq7bgrp9edj7y

**Comment by AayushChaudhary0001** [comment]
Tested and approved, LGTM

- This PR is tested and the setup of OAuth 2.0 is completely working fine.
- This PR was tested on Jira cloud latest version and MM version 7.8.2.
- During the testing, the notifications were also found working.

**Review by AayushChaudhary0001 — APPROVED** [review]

**Comment by mickmister** [comment]
> During the testing, the notifications were also found working.

@DHaussermann It seems the Jira comment events are fixed now. Merging the PR, and we can cover any more regression testing in the release testing.

**Pull request merged**